In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [4]:
gdp_df = pd.read_csv('../data/gdp_percapita.csv')

In [9]:
#looking at first 5 rows of df
gdp_df.head()

Country or Area  Year        Value  Value Footnotes
0     Afghanistan  2021  1517.016266              NaN
1     Afghanistan  2020  1968.341002              NaN
2     Afghanistan  2019  2079.921861              NaN
3     Afghanistan  2018  2060.698973              NaN
4     Afghanistan  2017  2096.093111              NaN

In [10]:
#looking at bottom 5 rows of df
gdp_df.tail()

Country or Area  Year        Value  Value Footnotes
7657        Zimbabwe  1994  2670.106615              NaN
7658        Zimbabwe  1993  2458.783255              NaN
7659        Zimbabwe  1992  2468.278257              NaN
7660        Zimbabwe  1991  2781.787843              NaN
7661        Zimbabwe  1990  2704.757299              NaN

## Q5: How many rows and columns are in gdp_df? What are the data types of each column?

In [8]:
gdp_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7662 entries, 0 to 7661
Data columns (total 4 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Country or Area  7662 non-null   object 
 1   Year             7662 non-null   int64  
 2   Value            7662 non-null   float64
 3   Value Footnotes  0 non-null      float64
dtypes: float64(2), int64(1), object(1)
memory usage: 239.6+ KB


## answer: 
7662 rows, 4 columns: 
country/area - object; 
year - integer; 
value - float; 
value footnotes - float

## Q6: Drop the Value Footnotes column and rename the remaining three to 'Country', 'Year', and 'GDP_Per_Capita'.

In [15]:
#df.drop("age", axis='columns') -- documentation for dropping a column
#gdp_df = gdp_df.drop('Value Footnotes', axis = 'columns')
gdp_df.head()

Country or Area  Year        Value
0     Afghanistan  2021  1517.016266
1     Afghanistan  2020  1968.341002
2     Afghanistan  2019  2079.921861
3     Afghanistan  2018  2060.698973
4     Afghanistan  2017  2096.093111

In [19]:
#renaming remaining columns

gdp_df = gdp_df.rename(columns = {'Country or Area': 'Country', 'Year':'Year', 'Value': 'GDP_Per_Capita'})

In [20]:
gdp_df.head()

Country  Year  GDP_Per_Capita
0  Afghanistan  2021     1517.016266
1  Afghanistan  2020     1968.341002
2  Afghanistan  2019     2079.921861
3  Afghanistan  2018     2060.698973
4  Afghanistan  2017     2096.093111

## answer: see above

## Q7: How many countries have data for all years? Which countries are missing many years of data? Look at the number of observations per year. What do you notice?

In [24]:
#unique values for year column
gdp_df['Year'].unique()
# 33 years total

array([2021, 2020, 2019, 2018, 2017, 2016, 2015, 2014, 2013, 2012, 2011,
       2010, 2009, 2008, 2007, 2006, 2005, 2004, 2003, 2002, 2022, 2001,
       2000, 1999, 1998, 1997, 1996, 1995, 1994, 1993, 1992, 1991, 1990],
      dtype=int64)

In [25]:
gdp_df['Country'].value_counts()
# 242 countries total

Country
Least developed countries: UN classification    33
Middle East & North Africa                      33
Middle East & North Africa (IDA & IBRD)         33
Middle income                                   33
Mongolia                                        33
                                                ..
Kosovo                                          15
Sint Maarten (Dutch part)                       14
Turks and Caicos Islands                        12
Somalia                                         10
Djibouti                                        10
Name: count, Length: 242, dtype: int64

In [38]:
grouped = gdp_df.groupby('Country')['Year'].count()
grouped

Country
Afghanistan                    20
Africa Eastern and Southern    33
Africa Western and Central     33
Albania                        33
Algeria                        33
                               ..
Viet Nam                       33
West Bank and Gaza             29
World                          33
Zambia                         33
Zimbabwe                       33
Name: Year, Length: 242, dtype: int64

In [42]:
#grouped = df.groupby('mygroups').sum().reset_index()
#grouped.sort_values('mygroups', ascending=False)

group_new = gdp_df.groupby('Country')['Year'].count().reset_index()
group_new.sort_values('Year', ascending=False)

Country  Year
121       Least developed countries: UN classification    33
143                         Middle East & North Africa    33
145  Middle East & North Africa (excluding high inc...    33
146                                      Middle income    33
148                                           Mongolia    33
..                                                 ...   ...
112                                             Kosovo    15
192                          Sint Maarten (Dutch part)    14
225                           Turks and Caicos Islands    12
197                                            Somalia    10
53                                            Djibouti    10

[242 rows x 2 columns]

In [51]:
group_new[group_new['Year'] < 33]
# list of countries missing data

Country  Year
0                  Afghanistan    20
10                       Aruba    32
22                      Bhutan    32
24      Bosnia and Herzegovina    29
32                    Cambodia    30
36              Cayman Islands    16
46                     Croatia    28
47                    Curaçao    22
53                    Djibouti    10
64                     Estonia    28
91                     Hungary    32
97                     Iceland    28
103                     Israel    28
112                     Kosovo    15
113                     Kuwait    31
120                     Latvia    28
122                    Lebanon    32
124                    Liberia    23
125                      Libya    24
126                  Lithuania    28
135                   Maldives    28
147                    Moldova    28
149                 Montenegro    26
154                      Nauru    19
166         Other small states    23
169                      Palau    22
180                      Qatar    23
185                 San Marino    25
188                     Serbia    28
192  Sint Maarten (Dutch part)    14
193            Slovak Republic    31
194                   Slovenia    28
195               Small states    26
197                    Somalia    10
213    São Tomé and Principe    22
219                Timor-Leste    23
221                      Tonga    32
224               Turkmenistan    31
225   Turks and Caicos Islands    12
238         West Bank and Gaza    29

In [48]:
group_new[group_new['Year'] < 33].count()
# of countries missing data = 40
# from line 25 above, there are 242 countries total
# 242 - 40 = 202 countries with all years of data

Country    40
Year       40
dtype: int64

In [26]:
gdp_df['Year'].value_counts()

Year
2013    242
2016    242
2014    242
2015    242
2020    242
2017    242
2018    242
2019    242
2021    241
2012    240
2011    240
2010    239
2009    239
2008    238
2007    237
2006    237
2004    236
2005    236
2003    235
2002    235
2001    234
2000    233
2022    232
1999    227
1998    226
1997    226
1996    223
1995    223
1994    213
1993    211
1992    210
1991    208
1990    207
Name: count, dtype: int64

In [27]:
gdp_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7662 entries, 0 to 7661
Data columns (total 3 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Country         7662 non-null   object 
 1   Year            7662 non-null   int64  
 2   GDP_Per_Capita  7662 non-null   float64
dtypes: float64(1), int64(1), object(1)
memory usage: 179.7+ KB


In [29]:
# of null values - none
gdp_df.isnull().sum()

Country           0
Year              0
GDP_Per_Capita    0
dtype: int64